crux_core/src/steps.rs

```diff
@@ -1,143 +1,88 @@
-use std::{
-    collections::{hash_map::Entry, HashMap},
-    fmt,
-    sync::Mutex,
-};
+use std::fmt::{self, Debug};
+use thiserror::Error;
 
-use uuid::Uuid;
+use serde::Serialize;
 
-use crate::Request;
+use crate::capability::Operation;
 
-pub(crate) struct Step<T> {
-    pub(crate) payload: T, // T is an Operation first, then Effect once mapped
-    pub(crate) resolve: Option<Resolve>,
-}
+type ResolveOnce<Out> = Box<dyn FnOnce(Out) + Send>;
+type ResolveMany<Out> = Box<dyn Fn(Out) -> Result<(), ()> + Send>;
 
-type OnceFn = dyn FnOnce(&[u8]) + Send;
-type ManyFn = dyn Fn(&[u8]) -> Result<(), ()> + Send;
+#[derive(Serialize)]
+pub struct Step<Op>(pub Op, #[serde(skip)] pub Option<Resolve<Op::Output>>)
+where
+    Op: Operation;
 
-pub(crate) enum Resolve {
-    Once(Box<OnceFn>),
-    Many(Box<ManyFn>),
+pub enum Resolve<Out> {
+    Once(ResolveOnce<Out>),
+    Many(ResolveMany<Out>),
 }
 
-impl<T> Step<T> {
-    pub fn resolves_never(payload: T) -> Self {
-        Self {
-            payload,
-            resolve: None,
-        }
+impl<Op> Step<Op>
+where
+    Op: Operation,
+{
+    pub(crate) fn resolves_never(payload: Op) -> Self {
+        Self(payload, None)
     }
 
-    pub fn resolves_once<F>(payload: T, resume: F) -> Self
+    pub(crate) fn resolves_once<F>(payload: Op, resolve: F) -> Self
     where
-        F: FnOnce(&[u8]) + Send + 'static,
+        F: FnOnce(Op::Output) + Send + 'static,
     {
-        Self {
-            payload,
-            resolve: Some(Resolve::Once(Box::new(resume))),
-        }
+        Self(payload, Some(Resolve::Once(Box::new(resolve))))
     }
 
-    pub fn resolves_many_times<F>(payload: T, resume: F) -> Self
+    pub(crate) fn resolves_many_times<F>(payload: Op, resolve: F) -> Self
     where
-        F: Fn(&[u8]) -> Result<(), ()> + Send + 'static,
+        F: Fn(Op::Output) -> Result<(), ()> + Send + 'static,
     {
-        Self {
-            payload,
-            resolve: Some(Resolve::Many(Box::new(resume))),
-        }
+        Self(payload, Some(Resolve::Many(Box::new(resolve))))
     }
 
-    pub fn map_effect<Ef, F>(self, f: F) -> Step<Ef>
-    where
-        F: Fn(T) -> Ef + Sync + Send + Copy + 'static,
-        T: 'static,
-        Ef: 'static,
-    {
-        Step {
-            payload: f(self.payload),
-            resolve: self.resolve,
+    pub fn resolve(&mut self, output: Op::Output) -> Result<(), ResolveError> {
```

Comment:
Actually needs to be `pub(crate)`, because the capability runtime needs to execute after each step is resolved - we can't let this be called from outside Crux unless we find a way to wake the runtime app right after.

There may be a way to do that with the wakers – `resolve` is basically a waker on the future that produced the step.